Have data frames handled appropriately by parallize

dputler opened the pull request:
Shivaram,
The issues we discussed concerning a data frame returning returning the value of TRUE when it is an argument to is.list().

shivaram commented on pkg/R/context.R:
```diff
@@ -85,11 +85,9 @@ parallelize <- function(sc, coll, numSlices = 1) {
-      message(paste("context.R: parallelize() currently only supports lists and vectors.",
-                    "Calling as.list() to coerce coll into a list."))
-    }
+  if ((!is.list(coll) && !is.vector(coll)) || is.data.frame(col)) {
```
This should be `is.data.frame(coll)` (i.e. the variable has two `l`s)

dputler replied:
Sorry, let me fix and resubmit.

 

 

 

From: Shivaram Venkataraman [mailto:notifications@github.com] 
Sent: Sunday, February 01, 2015 9:58 PM
To: amplab-extras/SparkR-pkg
Cc: dputler
Subject: Re: [SparkR-pkg] Have data frames handled appropriately by parallize (#150)

 

In pkg/R/context.R <https://github.com/amplab-extras/SparkR-pkg/pull/150#discussion_r23908070> :

> @@ -85,11 +85,9 @@ parallelize <- function(sc, coll, numSlices = 1) {
>    # TODO: bound/safeguard numSlices
>    # TODO: unit tests for if the split works for all primitives
>    # TODO: support matrix, data frame, etc
> -  if (!is.list(coll)) {
> -    if (!is.vector(coll)) {
> -      message(paste("context.R: parallelize() currently only supports lists and vectors.",
> -                    "Calling as.list() to coerce coll into a list."))
> -    }
> +  if ((!is.list(coll) && !is.vector(coll)) || is.data.frame(col)) {

This should be is.data.frame(coll) (i.e. the variable has two ls)

—
Reply to this email directly or view it on GitHub <https://github.com/amplab-extras/SparkR-pkg/pull/150/files#r23908070> .  <https://github.com/notifications/beacon/AIGvMg2qlEW-YLsBwG8-PJQvVfbeZ1Tbks5nnwlogaJpZM4DaK_g.gif>

sun-rui commented:
as.list() on a data frame converts it by columns. Does it make more sense to convert a data frame by rows?

dputler commented:
You have a point about this. One thing that complicates the situation is that parallelize currently creates a RDD, and a row based approach is really more appropriate for a SchemaRDD. We actually have in the works a write.SchemaRDD function that writes an R data frame to a SchemaRDD (each row is a Row RDD). The function has a numSlices option which partitions the SchemaRDD by rows.

 

Currently, when parallelize is used with a data frame, the return value from is.list() is TRUE, so the function attempts to directly parallelize the data frame, which results in the rows being sliced, but not in the expected way, and the functions frequently issues a warning. The current intended behavior of the function is to turn everything that isn’t a vector or a list into a list, which for a data frame handles things by columns. Because of how is.list() handles data frames, the function doesn’t produce the anticipated behavior. My proposed “fix” does, but whether that is really the desired behavior is another matter.

 

I’m going to hold-off on cleaning up my pull request until there is an agreement on the desired behavior. A related point is that currently if an R matrix is given as an argument to parallelize, as.list converts it into a list that consists of a set of single data elements, and since it there is no meta data conveyed, there isn’t really a clean way to reconstruct the original matrix once it is in Spark.

 

From: sun-rui [mailto:notifications@github.com] 
Sent: Monday, February 02, 2015 12:47 AM
To: amplab-extras/SparkR-pkg
Cc: dputler
Subject: Re: [SparkR-pkg] Have data frames handled appropriately by parallize (#150)

 

as.list() on a data frame converts it by columns. Does it make more sense to convert a data frame by rows?

—
Reply to this email directly or view it on GitHub <https://github.com/amplab-extras/SparkR-pkg/pull/150#issuecomment-72422857> .  <https://github.com/notifications/beacon/AIGvMt9r3mT9pPMJWK4G-m8jD-dJu2z2ks5nnzD6gaJpZM4DaK_g.gif>

shivaram commented:
Hmm this is an interesting question -- Here is my take on it
- In the long run we should of course make `parallelize` create a SchemaRDD / DataFrame as Dan suggested.
- In the meanwhile, we could try and special case the data.frame and matrix cases to work more reasonably. For example, we could split the data.frame row-wise such that every partition is itself a data.frame and one can run `lapplyPartition` on it. I think we can do this if we call `split` with `nrow(coll)` instead of `length(coll)` in the same function. And we could of course try something like this for matrices.

However, I am not sure the benefits from doing this will be more than the possible confusion it causes users. Because if we expect `lapply` to work on RDDs as it works locally, then I think `as.list` is the best we can do ?

piccolbo commented:
The short term version is how plyrmr does it as well. For data frames use
split.data.frame and for matrices, surprisingly, split.data.frame.
split.matrix does something else. But I concur with Shivaram that adding
this behavior to lapply, which means list-apply, is not a great idea.
lapply is used all the time to iterate over the columns, e.g.
lapply(mtcars, class). Unfortunately, there isn't really an apply family
function for data frames, that's why plyr and dplyr are so popular, besides
other reasons. The apply function can run on data.frames, but it will
return a matrix.


On Mon, Feb 2, 2015 at 2:16 PM, Shivaram Venkataraman <
notifications@github.com> wrote:

> Hmm this is an interesting question -- Here is my take on it
>
>    - In the long run we should of course make parallelize create a
>    SchemaRDD / DataFrame as Dan suggested.
>    - In the meanwhile, we could try and special case the data.frame and
>    matrix cases to work more reasonably. For example, we could split the
>    data.frame row-wise such that every partition is itself a data.frame and
>    one can run lapplyPartition on it. I think we can do this if we call
>    split with nrow(coll) instead of length(coll) in the same function.
>    And we could of course try something like this for matrices.
>
> However, I am not sure the benefits from doing this will be more than the
> possible confusion it causes users. Because if we expect lapply to work
> on RDDs as it works locally, then I think as.list is the best we can do ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/amplab-extras/SparkR-pkg/pull/150#issuecomment-72550099>
> .
>

dputler commented:
If the goal is to parallelize things so that when lapply is called against them, the behavior matches traditional R, then everything that is not either a vector or a list should be coerced to being a list before they are parallelized to make their behavior consistent with traditional R.

 

Replicating plyrmr functionality seems very worthwhile, but seems like a slightly different issue.

 

My two cents.

 

From: Antonio Piccolboni [mailto:notifications@github.com] 
Sent: Monday, February 02, 2015 2:35 PM
To: amplab-extras/SparkR-pkg
Cc: dputler
Subject: Re: [SparkR-pkg] Have data frames handled appropriately by parallize (#150)

 

The short term version is how plyrmr does it as well. For data frames use 
split.data.frame and for matrices, surprisingly, split.data.frame. 
split.matrix does something else. But I concur with Shivaram that adding 
this behavior to lapply, which means list-apply, is not a great idea. 
lapply is used all the time to iterate over the columns, e.g. 
lapply(mtcars, class). Unfortunately, there isn't really an apply family 
function for data frames, that's why plyr and dplyr are so popular, besides 
other reasons. The apply function can run on data.frames, but it will 
return a matrix. 


On Mon, Feb 2, 2015 at 2:16 PM, Shivaram Venkataraman < 
notifications@github.com <mailto:notifications@github.com> > wrote: 

> Hmm this is an interesting question -- Here is my take on it 
> 
> - In the long run we should of course make parallelize create a 
> SchemaRDD / DataFrame as Dan suggested. 
> - In the meanwhile, we could try and special case the data.frame and 
> matrix cases to work more reasonably. For example, we could split the 
> data.frame row-wise such that every partition is itself a data.frame and 
> one can run lapplyPartition on it. I think we can do this if we call 
> split with nrow(coll) instead of length(coll) in the same function. 
> And we could of course try something like this for matrices. 
> 
> However, I am not sure the benefits from doing this will be more than the 
> possible confusion it causes users. Because if we expect lapply to work 
> on RDDs as it works locally, then I think as.list is the best we can do ? 
> 
> — 
> Reply to this email directly or view it on GitHub 
> <https://github.com/amplab-extras/SparkR-pkg/pull/150#issuecomment-72550099> 
> . 
> 

—
Reply to this email directly or view it on GitHub <https://github.com/amplab-extras/SparkR-pkg/pull/150#issuecomment-72553187> .  <https://github.com/notifications/beacon/AIGvMpF3e5tEJakhwxp_dT0KEyKoehWwks5nn_MhgaJpZM4DaK_g.gif>

piccolbo commented:
Replicating plyrmr seems a very bad idea, I hope I didn't even remotely
suggest it. It works on top of sparkR and can share contexts and data. I
don't see the point in copying it, just install and load.

On Mon, Feb 2, 2015 at 2:52 PM, dputler <notifications@github.com> wrote:

> If the goal is to parallelize things so that when lapply is called against
> them, the behavior matches traditional R, then everything that is not
> either a vector or a list should be coerced to being a list before they are
> parallelized to make their behavior consistent with traditional R.
>
>
>
> Replicating plyrmr functionality seems very worthwhile, but seems like a
> slightly different issue.
>
>
>
> My two cents.
>
>
>
> From: Antonio Piccolboni [mailto:notifications@github.com]
> Sent: Monday, February 02, 2015 2:35 PM
> To: amplab-extras/SparkR-pkg
> Cc: dputler
> Subject: Re: [SparkR-pkg] Have data frames handled appropriately by
> parallize (#150)
>
>
>
> The short term version is how plyrmr does it as well. For data frames use
> split.data.frame and for matrices, surprisingly, split.data.frame.
> split.matrix does something else. But I concur with Shivaram that adding
> this behavior to lapply, which means list-apply, is not a great idea.
> lapply is used all the time to iterate over the columns, e.g.
> lapply(mtcars, class). Unfortunately, there isn't really an apply family
> function for data frames, that's why plyr and dplyr are so popular,
> besides
> other reasons. The apply function can run on data.frames, but it will
> return a matrix.
>
>
> On Mon, Feb 2, 2015 at 2:16 PM, Shivaram Venkataraman <
> notifications@github.com <mailto:notifications@github.com> > wrote:
>
> > Hmm this is an interesting question -- Here is my take on it
> >
> > - In the long run we should of course make parallelize create a
> > SchemaRDD / DataFrame as Dan suggested.
> > - In the meanwhile, we could try and special case the data.frame and
> > matrix cases to work more reasonably. For example, we could split the
> > data.frame row-wise such that every partition is itself a data.frame and
> > one can run lapplyPartition on it. I think we can do this if we call
> > split with nrow(coll) instead of length(coll) in the same function.
> > And we could of course try something like this for matrices.
> >
> > However, I am not sure the benefits from doing this will be more than
> the
> > possible confusion it causes users. Because if we expect lapply to work
> > on RDDs as it works locally, then I think as.list is the best we can do
> ?
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <
> https://github.com/amplab-extras/SparkR-pkg/pull/150#issuecomment-72550099>
>
> > .
> >
>
> —
> Reply to this email directly or view it on GitHub <
> https://github.com/amplab-extras/SparkR-pkg/pull/150#issuecomment-72553187>
> . <
> https://github.com/notifications/beacon/AIGvMpF3e5tEJakhwxp_dT0KEyKoehWwks5nn_MhgaJpZM4DaK_g.gif>
>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/amplab-extras/SparkR-pkg/pull/150#issuecomment-72556084>
> .
>

shivaram commented:
Yeah I don't mind the coercing everything to a list as a short-term solution. We could also print a warning if we detect a data-frame saying that `NOTE: DataFrames will be get parallelized by columns` or something like that.

dputler commented:
I can do that.

 

One R factoid that can be used here is that is.list() returns TRUE for a list and a data frame, it turns out that is.vector() returns TRUE for a list, but FALSE for a data frame. So, the actual test just needs to be based on a call to is.vector().

 

From: Shivaram Venkataraman [mailto:notifications@github.com] 
Sent: Monday, February 02, 2015 3:27 PM
To: amplab-extras/SparkR-pkg
Cc: dputler
Subject: Re: [SparkR-pkg] Have data frames handled appropriately by parallize (#150)

 

Yeah I don't mind the coercing everything to a list as a short-term solution. We could also print a warning if we detect a data-frame saying that NOTE: DataFrames will be get parallelized by columns or something like that. 

—
Reply to this email directly or view it on GitHub <https://github.com/amplab-extras/SparkR-pkg/pull/150#issuecomment-72561141> .  <https://github.com/notifications/beacon/AIGvMvTHNfo2pvkmuiZQDrGBXSG5ra64ks5nn_9cgaJpZM4DaK_g.gif>

piccolbo commented:
I wouldn't be so sure

> test = structure(list(data = 1), att = 1)
> is.vector(test)
[1] FALSE
> is.list(test)
[1] TRUE

On Mon, Feb 2, 2015 at 4:08 PM, dputler <notifications@github.com> wrote:

> I can do that.
>
>
>
> One R factoid that can be used here is that is.list() returns TRUE for a
> list and a data frame, it turns out that is.vector() returns TRUE for a
> list, but FALSE for a data frame. So, the actual test just needs to be
> based on a call to is.vector().
>
>
>
> From: Shivaram Venkataraman [mailto:notifications@github.com]
> Sent: Monday, February 02, 2015 3:27 PM
> To: amplab-extras/SparkR-pkg
> Cc: dputler
> Subject: Re: [SparkR-pkg] Have data frames handled appropriately by
> parallize (#150)
>
>
>
> Yeah I don't mind the coercing everything to a list as a short-term
> solution. We could also print a warning if we detect a data-frame saying
> that NOTE: DataFrames will be get parallelized by columns or something like
> that.
>
> —
> Reply to this email directly or view it on GitHub <
> https://github.com/amplab-extras/SparkR-pkg/pull/150#issuecomment-72561141>
> . <
> https://github.com/notifications/beacon/AIGvMvTHNfo2pvkmuiZQDrGBXSG5ra64ks5nn_9cgaJpZM4DaK_g.gif>
>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/amplab-extras/SparkR-pkg/pull/150#issuecomment-72566257>
> .
>

dputler commented:
Thanks, it would appear that the addition of the attribute causes a reversal in the behavior of is.vector(), which I didn’t anticipate.

 

From: Antonio Piccolboni [mailto:notifications@github.com] 
Sent: Monday, February 02, 2015 4:13 PM
To: amplab-extras/SparkR-pkg
Cc: dputler
Subject: Re: [SparkR-pkg] Have data frames handled appropriately by parallize (#150)

 

I wouldn't be so sure 

> test = structure(list(data = 1), att = 1) 
> is.vector(test) 
[1] FALSE 
> is.list(test) 
[1] TRUE 

On Mon, Feb 2, 2015 at 4:08 PM, dputler <notifications@github.com <mailto:notifications@github.com> > wrote: 

> I can do that. 
> 
> 
> 
> One R factoid that can be used here is that is.list() returns TRUE for a 
> list and a data frame, it turns out that is.vector() returns TRUE for a 
> list, but FALSE for a data frame. So, the actual test just needs to be 
> based on a call to is.vector(). 
> 
> 
> 
> From: Shivaram Venkataraman [mailto:notifications@github.com] 
> Sent: Monday, February 02, 2015 3:27 PM 
> To: amplab-extras/SparkR-pkg 
> Cc: dputler 
> Subject: Re: [SparkR-pkg] Have data frames handled appropriately by 
> parallize (#150) 
> 
> 
> 
> Yeah I don't mind the coercing everything to a list as a short-term 
> solution. We could also print a warning if we detect a data-frame saying 
> that NOTE: DataFrames will be get parallelized by columns or something like 
> that. 
> 
> — 
> Reply to this email directly or view it on GitHub < 
> https://github.com/amplab-extras/SparkR-pkg/pull/150#issuecomment-72561141> 
> . < 
> https://github.com/notifications/beacon/AIGvMvTHNfo2pvkmuiZQDrGBXSG5ra64ks5nn_9cgaJpZM4DaK_g.gif> 
> 
> 
> — 
> Reply to this email directly or view it on GitHub 
> <https://github.com/amplab-extras/SparkR-pkg/pull/150#issuecomment-72566257> 
> . 
> 

—
Reply to this email directly or view it on GitHub <https://github.com/amplab-extras/SparkR-pkg/pull/150#issuecomment-72566794> .  <https://github.com/notifications/beacon/AIGvMmIYRF31n_Re2AXLUnY2Atr8Qxzmks5noAoJgaJpZM4DaK_g.gif>
